stew/eh.nim

```diff
@@ -0,0 +1,389 @@
+
+    # Change the original proc name
+    procDef.name = genSym(nskProc, $procDef.name)
+
+    var wrapperBody = newNimNode(nnkStmtList, procDef.body)
+    if OrigResultType.kind == nnkEmpty or eqIdent(OrigResultType, "void"):
+      RaisingType = newTree(nnkBracketExpr, ident"Raising",
+                            ErrorsTuple, bindSym"VoidResult")
+      wrapperBody.add(
+        genForwardingCall(procDef),
+        newCall(RaisingType, newTree(nnkObjConstr, bindSym"VoidResult")))
+    else:
+      RaisingType = newTree(nnkBracketExpr, ident"Raising",
+                            ErrorsTuple, OrigResultType)
+      wrapperBody.add newCall(RaisingType, genForwardingCall(procDef))
+
+    wrapperDef.params[0] = if generateTemplate: ident"untyped"
+                           else: RaisingType
+    wrapperDef.body = wrapperBody
+
+    result = newStmtList(procDef, wrapperDef)
+  else:
+    result = procDef
+
+  storeMacroResult result
+
+macro checkForUnhandledErrors(origHandledErrors, raisedErrors: typed) =
+  # This macro is executed with two tuples:
+  #
+  # 1. The list of errors handled at the call-site which will
+  #    have a line info matching the call-site.
+  # 2. The list of errors that the called function is raising.
+  #    The lineinfo here points to the definition of the function.
+
+  # For accidental reasons, the first tuple will be recognized as a
+  # typedesc, while the second won't be (beware because this can be
+  # considered a bug in Nim):
+  var handledErrors = getTypeInst(origHandledErrors)
+  if handledErrors.kind == nnkBracketExpr:
+    handledErrors = handledErrors[1]
+
+  assert handledErrors.kind == nnkTupleConstr and
+         raisedErrors.kind == nnkTupleConstr
+
+  # Here, we'll store the list of errors that the user missed:
+  var unhandledErrors = newTree(nnkPar)
+
+  # We loop through the raised errors and check whether they have
+  # an appropriate handler:
+  for raised in raisedErrors:
+    block findHandler:
+      template tryFindingHandler(raisedType) =
+        for handled in handledErrors:
+          if sameType(raisedType, handled):
+            break findHandler
+
+      tryFindingHandler raised
+      # A base type of the raised exception may be handled instead
+      for baseType in raised.baseTypes:
+        tryFindingHandler baseType
+
+      unhandledErrors.add raised
+
+  if unhandledErrors.len > 0:
+    let errMsg = "The following errors are not handled: $1" % [unhandledErrors.repr]
+    error errMsg, origHandledErrors
+
+template raising*[E, R](x: Raising[E, R]): R =
+  ## `raising` is used to mark locations in the code that might
+  ## raise exceptions. It disarms the type-safety checks imposed
+  ## by the `errors` pragma.
+  distinctBase(x)
+
+template raising*[R, E](x: Result[R, E]): R =
+  tryGet(x)
+
+const
+  incorrectChkSyntaxMsg =
+    "The `check` handlers block should consist of `ExceptionType: Value/Block` pairs"
+
+template either*[E, R](x: Raising[E, R], otherwise: R): R =
```

Comment:
this is already called `get` for option and result..

Reply:
`get` is not lazy though. Also, with `either` you can have a `noReturn` statement in the otherwise branch:

```nim
let x = either foo():
               raise (ref MyException)(msg: "...")

```

Reply:
should we make it lazy? seems reasonable?

Reply:
The point is that we can't do it for the `Option` type, so I decided to use a new name. `either` was also reading better in English when you consider the semantics of the operation.

Reply:
https://github.com/status-im/nim-stew/blob/5512e89d4cfc56745840c74f301e19c43cf84e5c/stew/results.nim#L484 - thought I had something like that..

Reply:
though it calls get - hm, it should probably be rewritten

Reply:
which reminds me why I originally wrote `or` to take an untyped rhs:
```
template `or`*[T, E](self: Result[T, E], other: untyped): untyped ...

res or raise XXX
```

Reply:
but that doesn't work.. hmm.. not sure why, gives `expression expected found raise`

Reply:
It's the Nim parser. I wanted to use the `??` operator initially, but after realizing that the parser doesn't like `return` and `raise` used with infix operators, I've settled on `either`. `valueOr` is also OK as far as naming goes, but `either`is slightly more visually pleasing perhaps.

Reply:
is it fixable in the language? ie for a future sugary syntax, it would be a nice-to-have

Reply:
Maybe. It's hard to tell about these grammar changes before you try them out.